lib/puppet/file_serving/http_metadata.rb

```diff
@@ -0,0 +1,29 @@
+require 'puppet/file_serving/metadata'
+
+# Simplified metadata representation, suitable for the information
+# that is available from HTTP headers.
+class Puppet::FileServing::HttpMetadata < Puppet::FileServing::Metadata
+
+  def initialize(http_response, path = '/dev/null')
+    super(path)
+
+    # ignore options that do not apply to HTTP metadata
+    @owner = @group = @mode = nil
+
+    if checksum = http_response['content-md5']
+      # convert base64 digest to hex
+      checksum = checksum.unpack("m0").first.unpack("H*").first
+      @checksum_type = 'md5'
+      @checksum = "{md5}#{checksum}"
+    elsif mtime = DateTime.httpdate(http_response['last-modified']).to_time
+      @checksum_type = 'mtime'
+      @checksum = "{mtime}#{mtime}"
+    else
+      raise PuppetError, "HTTP response contained no usable checksum equivalent"
+    end
+
+    @ftype = 'file'
```

Comment:
I see what you mean now. But I believe you are mistaken.

This piece of code determines what the indirector reports to find upstream. For all intents and purposes, this is invariably a flat `file` here. Even if a manifest specifies `ensure => 'directory'`, the agent will sync a flat file. This is in keeping with the behavior when using `puppet://` or `file://` URLs.

Even if it would make sense to error out here, there is no way to catch that. The indirector must report the type of the entity it finds under the given key. It does not know about the specifics of the file resource on the other end.